### PR TITLE
docs: Add how-to for Python development (virtualenvs, make test)

### PR DESCRIPTION
**Pull request by timmc-edx**
We currently have "Getting Started" instructions in each repo's README, but they're terse and often out of date. By moving them to a central place, we can instead link to those docs and reduce our maintenance burden.

This would also involve updating OEP-55 to specify a link to this doc page rather than [suggesting an inlined Getting Started section](https://open-edx-proposals.readthedocs.io/en/latest/processes/oep-0055/decisions/0003-readme-specification.html).

### CODE REVIEW

File: source/developers/how-tos/get-ready-for-python-dev.rst
Comment (timmc-edx) [reply source/developers/how-tos/get-ready-for-python-dev.rst:71]: There's a huge amount of variation in what the various repos include. Maybe I should just make a list of targets that may or may not be relevant? I think those would be test, test-all, quality, validate, docs, diff_cover, coverage. Some even have targets for isort, pycodestyle, and pylint. I think in at least one, you're just supposed to run tox directly. (In some repos that only runs pytest, and other places it runs all the checks.)

It does seem like `make quality` is worth mentioning. Maybe nothing beyond that, and let CI checks take care of the rest?

My longer term goal is to homogenize the Makefile targets we use for Python. Writing this how-to really illustrates some of the current problems.

File: source/developers/how-tos/get-ready-for-python-dev.rst
Comment (sarina) [inline source/developers/how-tos/get-ready-for-python-dev.rst:79]: If you go through the steps of creating a branch then realize you need a fork, is it annoying at all? Or does it Just Work®?

File: source/developers/how-tos/get-ready-for-python-dev.rst
Comment (ashultz0) [reply source/developers/how-tos/get-ready-for-python-dev.rst:71]: Yeah I wouldn't make the list, that gets confusing. Make test and make quality work pretty often and running both of those will get you a long way to passing your PR checks.

File: source/developers/how-tos/get-ready-for-python-dev.rst
Comment (timmc-edx) [reply source/developers/how-tos/get-ready-for-python-dev.rst:79]: It should just work. Unless you check out a remote branch (creating a local one) or use `-u/--set-upstream` on a local branch, git doesn't have any particular opinion about what remote the branch goes with.

(I usually would use `git push -u fork` here, so that the branch always pushes to the same remote. Maybe I should include the `-u` option here?)

Comment (sarina) [reply source/developers/how-tos/get-ready-for-python-dev.rst:79]: I guess I wonder if it makes sense to instruct people to make a fork first

Comment (kdmccormick) [reply source/developers/how-tos/get-ready-for-python-dev.rst:79]: I think fork-by-default is a reasonable strategy that aligns with most open source projects on GitHub. I'm an org owner and I use forks for almost everything.

Only maintainers and release managers really need to push new branches upstream, and folks in those roles probably don't need a guide to do that.

Comment (timmc-edx) [reply source/developers/how-tos/get-ready-for-python-dev.rst:79]: Fair enough, I'll separate it out.

File: source/developers/how-tos/get-ready-for-python-dev.rst
Comment (feanil) [inline source/developers/how-tos/get-ready-for-python-dev.rst:48]: Maybe make this a `.. note::` directive and add a link to the virtualenv wrapper docs: https://virtualenvwrapper.readthedocs.io/en/stable/

File: source/developers/how-tos/get-ready-for-python-dev.rst
Comment (feanil) [reply source/developers/how-tos/get-ready-for-python-dev.rst:79]: +1 to assuming people are working with forks, the plan is to have reduced access for 2U in the future too so having the default docs assume you use forks first will be generally useful.

File: source/developers/how-tos/get-ready-for-python-dev.rst
Comment (kdmccormick) [inline source/developers/how-tos/get-ready-for-python-dev.rst:23]: I get what you're saying here, but from the perspective of a new Python dev who has no idea what special workflows are  (which IMO is the audience of this doc), this note amounts to saying:

> If these instructions conflict with the repository's instructions, then follow these instructions, except in the cases where you shouldn't, and in those cases you should follow repository's instructions.

which I think is more confusing and disheartening than it is helpful 😛

File: source/developers/how-tos/get-ready-for-python-dev.rst
Comment (kdmccormick) [inline source/developers/how-tos/get-ready-for-python-dev.rst:71]: Gah, I would love to standardize on something. I've become quite partial to tutor's way, which is to [namespace all the checks under `make test-*` and use `make test` as the omni-check](https://github.com/overhangio/tutor/blob/7d1a6fff6e71e8df4d159cf97bdca4d799dbe90e/Makefile#L29-L49), but I think I'd be happy with any standard at this point.

I know that's outside the scope. I think how you described the current situation here is great.

File: source/developers/how-tos/get-ready-for-python-dev.rst
Comment (kdmccormick) [inline source/developers/how-tos/get-ready-for-python-dev.rst:78]: This skips `git add`.

You could spell out `git add` as a step. Alternatively, you could be purposefully vague, and let the developer find a separate resource for learning git basics.
```suggestion
#. Run ``make test`` one more time and then commit your changes. Follow the `conventional commits`_ documentation. Make sure your commit message is informative and describes why the change is being made. While the first line of the message should be terse, the body of the message has plenty of room for details.
```

File: source/developers/how-tos/get-ready-for-python-dev.rst
Comment (kdmccormick) [inline source/developers/how-tos/get-ready-for-python-dev.rst:80]:
```suggestion
#. In GitHub, open a pull request (PR). In the PR description, include anything that could help reviewers understand and test your change.
#. If you have a repository maintainer in mind, you can reach out to them and ask for review. Otherwise, an Open edX project manager will triage your PR and find someone to review it. This process can take some time; we appreciate your patience!
```

"Ask for review" only works for folks who know maintainers/CCs. Without getting too deep into "How to make a PR", here are a couple items that could clue them into the next steps.

File: source/developers/how-tos/get-ready-for-python-dev.rst
Comment (nedbat) [inline source/developers/how-tos/get-ready-for-python-dev.rst:36]: I kind of understand why the `-3.8` suffix is there, but these instructions will only involve one venv, so maybe we can keep it simpler by skipping the suffix.

File: source/developers/how-tos/get-ready-for-python-dev.rst
Comment (kdmccormick) [reply source/developers/how-tos/get-ready-for-python-dev.rst:75]: FWIW I have been forcing myself to use switch and restore lately, and can confirm that they just make more sense.

File: source/developers/how-tos/get-ready-for-python-dev.rst
Comment (nedbat) [reply source/developers/how-tos/get-ready-for-python-dev.rst:46]: The Python cookiecutter doesn't mention it :(

File: source/developers/how-tos/get-ready-for-python-dev.rst
Comment (nedbat) [reply source/developers/how-tos/get-ready-for-python-dev.rst:79]: I haven't seen people using `fork` as the name of their personal forks.  More common is `origin` for your fork, and `upstream` for the upstream.

**Review by kdmccormick — APPROVED** [review]
I'm changing my review from Request Changes to Approve, just to be clear that none of my suggestions are meant to be blocking.

File: source/developers/how-tos/get-ready-for-python-dev.rst
Comment (timmc-edx) [reply source/developers/how-tos/get-ready-for-python-dev.rst:79]: I've seen both patterns. origin/upstream weirds me out because the upstream *is* the origin -- they more or less mean the same thing! Frankly, I use both, depending on whether I've initially cloned the authoritative repo or my fork. I'm too lazy to rename the remotes. :upside_down_face:

File: source/developers/how-tos/get-ready-for-python-dev.rst
Comment (timmc-edx) [reply source/developers/how-tos/get-ready-for-python-dev.rst:75]: I've never used it, and the man page has a big warning:

> THIS COMMAND IS EXPERIMENTAL. THE BEHAVIOR MAY CHANGE.

...but I'm guessing the basic behavior will remain unchanged? It looks like the main benefit is that you don't have to do `git stash; git checkout ...; git stash pop` when you're on the wrong branch at first.

Comment (nedbat) [reply source/developers/how-tos/get-ready-for-python-dev.rst:75]: I didn't realize there was a stash different with switch.  To me, the main advantage is that it's conceptually clearer.  I wish they would remove the EXPERIMENTAL warning. There isn't that much behavior to change in the first place.

File: source/developers/how-tos/get-ready-for-python-dev.rst
Comment (timmc-edx) [reply source/developers/how-tos/get-ready-for-python-dev.rst]: I've updated the PR to tell people to use a fork. It doesn't tell people how to _update_ their fork, but that might be out of scope.

Comment (sarina) [reply source/developers/how-tos/get-ready-for-python-dev.rst]: I think that is out of scope, but you could consider linking this: https://docs.openedx.org/en/latest/developers/quickstarts/first_openedx_pr.html

Comment (kdmccormick) [reply source/developers/how-tos/get-ready-for-python-dev.rst]: Thanks Tim, looks good.

> It doesn't tell people how to update their fork, but that might be out of scope.

I think that's fair. This doc probably isn't the right place for a comprehensive "how to use git with multiple remotes" guide.

**Comment by feanil** [comment]
@timmc-edx let me know when you think this is ready to merge and if you'd like me to squash it or retain all the commits.

File: source/developers/how-tos/get-ready-for-python-dev.rst
Comment (timmc-edx) [reply source/developers/how-tos/get-ready-for-python-dev.rst:6]: Yes, libraries too -- will change.

File: source/developers/how-tos/get-ready-for-python-dev.rst
Comment (timmc-edx) [reply source/developers/how-tos/get-ready-for-python-dev.rst:23]: It's tough to figure out how to say "a lot of those READMEs are wrong but some of them are not" in a useful way and that, frankly, isn't too embarrassing. Maybe this:

> These instructions will not work in all repositories. If they do not, check the repo's README for a Getting Started section for possible alternative instructions.

**Comment by timmc-edx** [comment]
I think I've addressed all feedback except Kyle's regarding the issue of conflicting instructions. Still pondering what to do there; suggestions very welcome.

File: source/developers/how-tos/get-ready-for-python-dev.rst
Comment (sarina) [reply source/developers/how-tos/get-ready-for-python-dev.rst:23]: I actually really like this wording!

**Comment by timmc-edx** [comment]
OK, I think it's ready to go.

**Comment by feanil** [comment]
@timmc-edx do you want to squash to something you're happy with or is it cool to just squash all the commits together upon merge?

**Comment by timmc-edx** [comment]
PR title & desc should be fine for the squashed commit message.

**Comment by feanil** [comment]
Thanks Tim! 🎉

**Pull request merged**
